### PR TITLE
[SPARK-30809][SQL] Review and fix issues in SQL API docs

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
- Add missing `since` annotation.
- Don't show classes under `org.apache.spark.sql.dynamicpruning` package in API docs.
- Fix the scope of `xxxExactNumeric` to remove it from the API docs.

### Why are the changes needed?
Avoid leaking APIs unintentionally in Spark 3.0.0.

### Does this PR introduce any user-facing change?
No. All these changes are to avoid leaking APIs unintentionally in Spark 3.0.0.

### How was this patch tested?
Manually generated the API docs and verified the above issues have been fixed.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Catalogs.java [hunk 1]
```diff
@@ -48,6 +48,7 @@ private Catalogs() {
    * @return an initialized CatalogPlugin
    * @throws CatalogNotFoundException if the plugin class cannot be found
    * @throws SparkException if the plugin class cannot be instantiated
+   * @since 3.0.0
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Catalogs.java]: This is a private class (`@Private`)

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/Transform.java [hunk 2]
```diff
@@ -24,6 +24,8 @@
  * <p>
  * For example, the transform date(ts) is used to derive a date value from a timestamp column. The
  * transform name is "date" and its argument is a reference to the "ts" column.
+ *
+ * @since 3.0.0
  */
 @Experimental
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/Transform.java]: I think all DS v2 classes are still evolving. Shall we use `@Evolving` consistently? e.g. https://github.com/apache/spark/pull/27560/files#diff-ab2a72871faeb3ad8bae1d7f951899deR29

cc @brkyvz @rdblue

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -834,6 +834,7 @@ object Unidoc {
       .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/util/collection")))
       .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/util/kvstore")))
       .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/catalyst")))
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/dynamicpruning")))
```
Comment (cloud-fan) [inline project/SparkBuild.scala]: shall we send a PR to move package first?

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 4]
```diff
@@ -108,7 +108,7 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
 /**
  * An exception thrown if the RPC is aborted.
  */
-class RpcAbortException(message: String) extends Exception(message)
+private[spark] class RpcAbortException(message: String) extends Exception(message)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: shall we just make the `rpc` package private?

File: project/SparkBuild.scala (hunk 3 above)
Comment (xuanyuanking) [reply project/SparkBuild.scala]: Sure, will do it today.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Catalogs.java (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Catalogs.java]: If it doesn't work, we can rewrite in with Scala and use `private[sql]`.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/ExternalCommandRunner.java [hunk 5]
```diff
@@ -27,6 +27,8 @@
  * cores for Solr and so on.
  * <p>
  * This interface will be instantiated when end users call `SparkSession#executeCommand`.
+ *
+ * @since 3.0.0
```
Comment (Ngone51) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/ExternalCommandRunner.java]: Oh..no...thanks for adding this. BTW, how do you find it?

File: project/SparkBuild.scala (hunk 3 above)
Comment (xuanyuanking) [reply project/SparkBuild.scala]: Done in #27581.

**Comment by cloud-fan** [comment]
any more updates?

**Comment by xuanyuanking** [comment]
Yes, will address comments today.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Catalogs.java (hunk 1 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Catalogs.java]: Thanks, rewrite done in 69ab608.

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala (hunk 4 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: Done in a42f54c.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/ExternalCommandRunner.java (hunk 5 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/ExternalCommandRunner.java]: I compare the latest Scala API doc against the 2.4 version. :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+   * This loads, instantiates, and initializes the catalog plugin for each call; it does not cache
+   * or reuse instances.
+   *
+   * @param name a String catalog name
+   * @param conf a SQLConf
+   * @return an initialized CatalogPlugin
+   * @throws CatalogNotFoundException if the plugin class cannot be found
+   * @throws SparkException           if the plugin class cannot be instantiated
+   */
+    @throws[CatalogNotFoundException]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: is the indentation corrected?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 7]
```diff
@@ -488,6 +488,8 @@ object DataType {
 
 /**
  * Jackson serializer for [[DataType]]. Internally this delegates to json4s based serialization.
+ *
+ * @since 3.0.0
  */
 class DataTypeJsonSerializer extends JsonSerializer[DataType] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: these 2 were added in https://github.com/apache/spark/pull/26127/files#diff-df78a74ef92d9b8fb4ac142ff9a62464R490

I don't think we want to publish them. Can we move them to a private package? like `xxx.catalyst.utils`
also cc @hvanhovell

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala [hunk 8]
```diff
@@ -22,7 +22,7 @@ import scala.math.Ordering
 
 import org.apache.spark.sql.types.Decimal.DecimalIsConflicted
 
-object ByteExactNumeric extends ByteIsIntegral with Ordering.ByteOrdering {
+private[sql] object ByteExactNumeric extends ByteIsIntegral with Ordering.ByteOrdering {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: I'm OK with it. But we should also move this file to a private package.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: we can move it in followup

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala (hunk 7 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: Good catch. We should definitely move it into catalyst. I am not sure if we can change the visibility (jackson might not like that), but we should just try.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: classes in the catalyst package are treated as private classes and we don't really need to add the `private[sql]`

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: Thanks, move them to `org.apache.spark.sql.catalyst.util.DataTypeJsonUtils`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala (hunk 6 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: Thanks, fix in 14dbf06.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala (hunk 8 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala]: Got it, others object in numerics.scala might also get moved together.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+   * @return an initialized CatalogPlugin
+   * @throws CatalogNotFoundException if the plugin class cannot be found
+   * @throws SparkException           if the plugin class cannot be instantiated
+   */
+  @throws[CatalogNotFoundException]
+  @throws[SparkException]
+  def load(name: String, conf: SQLConf): CatalogPlugin = {
+    var pluginClassName = ""
+    try {
+      pluginClassName = conf.getConfString("spark.sql.catalog." + name)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: we can make it more scala-ish:
```
val pluginClassName = try {
  ...
} ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    } catch {
+      case _: NoSuchElementException =>
+        throw new CatalogNotFoundException(
+          s"Catalog '$name' plugin class not found: spark.sql.catalog.$name is not defined")
+    }
+    val loader = Utils.getContextOrSparkClassLoader
+    try {
+      val pluginClass = loader.loadClass(pluginClassName)
+      if (!classOf[CatalogPlugin].isAssignableFrom(pluginClass)) throw new SparkException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: nit:
```
if (...) {
  throw ...
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        s"Plugin class for catalog '$name' does not implement CatalogPlugin: $pluginClassName")
+      val plugin = classOf[CatalogPlugin].cast(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: nit: `pluginClass.getDeclaredConstructor().newInstance().asInstanceOf[CatalogPlugin]`

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala (hunk 9 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: Thanks, done in 14b121e.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala (hunk 10 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: Thanks, done in 14b121e

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala (hunk 11 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala]: Thanks, done in 14b121e

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by xuanyuanking** [comment]
Thanks!

**Pull request closed without merging**